### ISSUE TITLE
Composite Res Type: Failed to create RefContentType for a public url

### ISSUE THREAD

**Issue by zhiyuli** [hydroshare/hydroshare#3175]
Both UI and REST API were failed to create RefContentType for url:
http://ameriflux.lbl.gov/sites/siteinfo/US-Rms

It is public and can be opened in browser without any issue. Not sure what was wrong.

**Comment by zhiyuli** [comment]
This ticket reported an issue different from #3156

**Comment by hyi** [comment]
@zhiyuli I debugged and confirmed this issue is specific to this URL, so it is not a regression. The ```urlopen()``` method for this url raised an HTTPError exception with 403: Forbidden. Also when you open this URL from web browser, the browser shows a ```not secure``` connection indication at the top URL bar. So this URL does not security issues that somehow failed urlopen() testing.

**Comment by zhiyuli** [comment]
@hyi 
Thanks for looking into this. 
Can we loose some security checks and restrictions imposed by urlopen() to allow the creation of RefFileType on more urls? This may help fix #3156 as well.

**Comment by hyi** [comment]
@zhiyuli I am looping in @horsburgh and @dtarb for them to chime in.

**Comment by zhiyuli** [comment]
@hyi 
Actually, I just found this url has a HTTPS version. Simply replace http with https:
https://ameriflux.lbl.gov/sites/siteinfo/US-CZ4

But I was not able to create RefFileType for it through UI or API, either.

**Comment by zhiyuli** [comment]
Browser says this https version is secured
![image](https://user-images.githubusercontent.com/10070599/53184674-b2b66d80-35ba-11e9-854b-b6adf3ad1925.png)

**Comment by horsburgh** [comment]
The HTTPS URL should work. Our intent was that any valid URL should work. If you look at the design document, we also acknowledged that some users might want to add URLs that require third party authentication and that is something the system should support. We decided that HydroShare would not try to do any sort of authentication for the user, but rather just pass them to the URL and let their browser handle it.

@hyi put validation in to make sure the user-input URL is valid. I'm not sure why the secure HTTPS URL doesn't work, but it should. With third party authentication, we are going to have to decide how to handle that on the validation of the user-input URL.

@zhiyuli - is there a reason you aren't using the DOI URLs for these pages? If I go to the link you listed and click on the DOI tab at the top, it takes me to a page with a URL-encoded DOI. Seems like those links would be way safer for referenced content.

**Comment by zhiyuli** [comment]
@zhiyuli 
Hmmm...good call! 
I didnt notice it has a DOI URL. The plain url is used by CZO database and we just migrate it to HS as is. But I can let CZO people aware of the DOI and maybe they would want to update their DB to use it.

**Comment by hyi** [comment]
@zhiyuli I have created a PR https://github.com/hydroshare/hydroshare/pull/3190 to fix this. The reason is that the urlopen() is blocked by this website as web spider/bot. To work around this, a web browser 'User-Agent' has to be added into request header and pass the request object to urlopen() rather than directly opening the url. I asked you to code review that PR. I also started the process to deploy the fix to park so that you can test and verify the fix there.

**Comment by hyi** [comment]
@zhiyuli park is deployed with the fix and I have verified that I am able to create a composite resource with this referenced url file now with the fix. The resource is here: https://park.hydroshare.org/resource/f7ac8f79b93449319bf8aee466355024/
However, I am not able to make this resource public. It always says the resource is missing keywords although I added keywords for sure. When I went back to view mode, added keywords don't show even though they showed up in edit mode. It appears to be a bug to me, but in case it is an operation error, I need someone to verify this is indeed an issue. Looping in @horsburgh and @Maurier. Steps to reproduce the issue I ran into:
create an empty composite resource with NO file uploaded, add a reference url into the resource, fill in abstract, and add subject/keywords, then try to make it public. I am not able to make it public.

**Comment by zhiyuli** [comment]
@hyi 
Confirmed: the url issue is fixed on park

@Maurier @horsburgh 
Confirmed: keywords didnt get saved; and I dont even see the "make public" icon

**Comment by horsburgh** [comment]
@hyi and @zhiyuli - I just created a new resource on www.hydroshare.org, added a reference to a website, keywords, abstract, etc. I saved it with no problems and made it public with no problems.

The buttons to make a resource public, private, etc. are no long on the page, they are in the manage access panel. Are you looking in the right place for the button?

I don't know what to say about the keywords not saving. It's working correctly on www.

**Comment by horsburgh** [comment]
@hyi - I was going to try on park, but I created a new account and never got the email to activate my account. I can't remember where I'm supposed to go to get that email.

**Comment by cuahsimarko** [comment]
This confused me for a while. I think you go to https://nginx1-beta.hydroshare.org:8443/webmail/src/webmail.php e-mailed you the pw

**Comment by horsburgh** [comment]
Something is different with respect to keywords on park versus www. I successfully created a new resource with keywords and referenced content and made it public on www. Trying to do the exact same things on park the keywords aren't saved, so I can't make it public.

But, you already knew this...

**Comment by hyi** [comment]
@horsburgh Not sure what happened on park either. As long as it works on www, we should be good. I used Jenkins to automatically deploy my branch to park, but since local_settings.py is not in the branch any more, automatic deploy failed initially and I had to ssh to park to create a local_settings.py and rerun database migration and collectstatic. Probably in the process something went awry although I just tried to run database migration and collectstatic again on park and it appeared everything got applied correctly without issues.
